apps/aecore/lib/aecore/peers/sync.ex

```diff
@@ -0,0 +1,607 @@
+  defp merge(old_hashes, []), do: old_hashes
+
+  defp merge([{{h_1, _hash_1}, _} | old_hashes], [{{h_2, hash_2}, map_2} | new_hashes])
+       when h_1 < h_2 do
+    merge(old_hashes, [{{h_2, hash_2}, map_2} | new_hashes])
+  end
+
+  defp merge([{{h_1, hash_1}, map_1} | old_hashes], [{{h_2, _hash_2}, _} | new_hashes])
+       when h_1 > h_2 do
+    merge([{{h_1, hash_1}, map_1} | old_hashes], new_hashes)
+  end
+
+  defp merge(old_hashes, [{{h_2, hash_2}, map_2} | new_hashes]) do
+    pick_same({{h_2, hash_2}, map_2}, old_hashes, new_hashes)
+  end
+
+  defp pick_same({{h, hash_2}, map_2}, [{{h, hash_1}, map_1} | old_hashes], new_hashes) do
+    case hash_1 == hash_2 do
+      true ->
+        [
+          {{h, hash_1}, Map.merge(map_1, map_2)}
+          | pick_same({{h, hash_2}, map_2}, old_hashes, new_hashes)
+        ]
+
+      false ->
+        [{{h, hash_1}, map_1} | pick_same({{h, hash_2}, map_2}, old_hashes, new_hashes)]
+    end
+  end
+
+  defp pick_same(_, old_hashes, new_hashes), do: merge(old_hashes, new_hashes)
+
+  defp fill_pool(peer_id, agreed_hash) do
+    case PeerConnection.get_n_successors(agreed_hash, @max_headers_per_chunk, peer_id) do
+      {:ok, []} ->
+        delete_from_pool(peer_id)
+        :done
+
+      {:ok, chunk_hashes} ->
+        hash_pool =
+          for chunk <- chunk_hashes do
+            {chunk, %{peer: peer_id}}
+          end
+
+        update_hash_pool(hash_pool)
+        {:filled_pool, length(chunk_hashes) - 1}
+
+      _err ->
+        delete_from_pool(peer_id)
+        {:error, :sync_abort}
+    end
+  end
+
+  # Check if we already have this block locally, is so
+  # take it from the chain
+  defp do_fetch_block(hash, peer_id) do
+    case Chain.get_block(hash) do
+      {:ok, block} ->
+        Logger.debug(fn -> "#{__MODULE__}: We already have this block!" end)
+        {:ok, false, block}
+
+      {:error, _} ->
+        do_fetch_block_ext(hash, peer_id)
+    end
+  end
+
+  # If we don't have the block locally, take it from the Remote Peer
+  defp do_fetch_block_ext(hash, peer_id) do
+    case PeerConnection.get_block(hash, peer_id) do
+      {:ok, block} ->
+        case BlockValidation.block_header_hash(block.header) === hash do
+          true ->
+            Logger.debug(fn ->
+              "#{__MODULE__}: Block #{inspect(block)} fetched from #{inspect(peer_id)}"
+            end)
+
+            {:ok, true, block}
+
+          false ->
+            {:error, :hash_mismatch}
+        end
+
+      err ->
+        Logger.debug(fn -> "#{__MODULE__}: Failed to fetch the block from #{inspect(peer_id)}" end)
+
+        err
+    end
+  end
+
+  # Try to fetch the pool of transactions
+  # from the Remote Peer we are connected to
+  defp do_fetch_mempool(peer_id) do
+    {:ok, pool} = PeerConnection.get_mempool(peer_id)
+    Logger.debug(fn -> "#{__MODULE__}: Mempool received from #{inspect(peer_id)}" end)
+    Enum.each(pool, fn {_hash, tx} -> Pool.add_transaction(tx) end)
```

Comment:
Heads up: get_mempool now returns a list of all transactions, not a map (should be changed to fn tx -> ...)